### PR TITLE
Add ParameterizedValue::IpAddress for postgres

### PR DESCRIPTION
**Pull request by tomhoule**

### CODE REVIEW

File: src/ast/values.rs [hunk 1]
```diff
@@ -265,6 +271,19 @@ impl<'a> ParameterizedValue<'a> {
         }
     }
 
+    #[cfg(feature = "postgresql")]
+    pub fn is_ip_address(&self) -> bool {
+        self.as_ip_address().is_some()
+    }
+
+    #[cfg(feature = "postgresql")]
+    pub fn as_ip_address(&self) -> Option<&std::net::IpAddr> {
```
Comment (pimeys) [inline src/ast/values.rs]: `IpAddr` is `Copy` so return by value.

File: src/connector/mysql/conversion.rs [hunk 2]
```diff
@@ -105,6 +105,8 @@ impl<'a> From<ParameterizedValue<'a>> for MyValue {
                 dt.second() as u8,
                 dt.timestamp_subsec_micros(),
             ),
+            #[cfg(feature = "postgresql")]
+            ParameterizedValue::IpAddress(ip_addr) => MyValue::Bytes(ip_addr.to_string().into_bytes()),
```
Comment (pimeys) [inline src/connector/mysql/conversion.rs]: hmm...

?

Comment (pimeys) [reply src/connector/mysql/conversion.rs]: I know this is kind of needed if you have both features enabled. But maybe the enum variant should not be under a feature flag then...

File: src/connector/sqlite/conversion.rs [hunk 3]
```diff
@@ -68,6 +68,9 @@ impl<'a> ToSql for ParameterizedValue<'a> {
             ParameterizedValue::Uuid(value) => ToSqlOutput::from(value.to_hyphenated().to_string()),
             #[cfg(feature = "chrono-0_4")]
             ParameterizedValue::DateTime(value) => ToSqlOutput::from(value.timestamp_millis()),
+
+            #[cfg(feature = "postgresql")]
+            ParameterizedValue::IpAddress(ip_addr) => ToSqlOutput::from(ip_addr.to_string().into_bytes()),
```
Comment (pimeys) [inline src/connector/sqlite/conversion.rs]: Same here. maybe do not hide the variant under a feature flag if all database connectors still need to react to the values if postgres is enabled.

**Comment by pimeys** [comment]
Also what bothers me a bit is that with PostgreSQL when we read an IP address value from the database, it'll be an IP address. If then we read an IP address from MySQL or SQLite, it will not be an IP address variant from ParameterizedValue.

This smells like a problem with our architecture with parameterized values...

**Comment by tomhoule** [comment]
Yes, maybe a trait-based type mapping approach like `tokio-postgres` would be more flexible than the enum.

**Comment by pimeys** [comment]
Do you have an idea how that would look like in a multi-database context such as Quaint?

**Comment by tomhoule** [comment]
We need to think really hard about this, the traits would have to be very generic.

Also we would probably have `Quaint<Postgres>`, `Quaint<Sqlite>` and so on to allow some types of parameters, etc. Then that type parameter would probably end up everywhere in the AST, it would be messy. But possible more powerful than an enum.

**Comment by pimeys** [comment]
Yeah... I don't really like how we have database-specific enum variants now and especially how you cannot map them from and back from the database so that they'll always turn into the same variant.

**Comment by pimeys** [comment]
Remember also how `ParameterizedValue` is the singular enum that will be parameterized out from the SQL queries. That role is extremely important, but maybe we should not return `ParameterizedValue`s in the `ResultSet`s.

**Comment by tomhoule** [comment]
Closing this, since we went with not extending `ParameterizedValue` for now.

**Comment by tomhoule** [comment]
Closing this for now, since we need to decide on an approach to have a rich type system for result sets in the future.

**Pull request closed without merging**